File: extension/src/list-diff.ts
/** List-section diff helpers for Candidate Review UI. */

import type { CandidateDetail, CandidateDiff, CandidateSummary } from "./types.js";

export type ListDiff = {
  added: string[];
  removed: string[];
  unchanged: string[];
  unchangedCount?: number;
  operation?: string;
};

export const LIST_DIFF_SECTIONS = [
  "important_terms",
  "tags",
  "keywords",
  "interests",
  "focus_terms",
] as const;

export function isListDiffSection(section: string): boolean {
  return (LIST_DIFF_SECTIONS as readonly string[]).includes(section);
}

export function diffStringList(existing: string[], proposed: string[]): ListDiff {
  const normalize = (s: string) => s.trim();
  const normKey = (s: string) => normalize(s).toLowerCase();

  const existingMap = new Map<string, string>();
  for (const v of existing) {
    const n = normalize(v);
    if (n) existingMap.set(normKey(n), n);
  }
  const proposedMap = new Map<string, string>();
  for (const v of proposed) {
    const n = normalize(v);
    if (n) proposedMap.set(normKey(n), n);
  }

  const added: string[] = [];
  const removed: string[] = [];
  const unchanged: string[] = [];

  for (const [key, value] of proposedMap) {
    if (existingMap.has(key)) unchanged.push(value);
    else added.push(value);
  }
  for (const [key, value] of existingMap) {
    if (!proposedMap.has(key)) removed.push(value);
  }

  added.sort((a, b) => a.localeCompare(b));
  removed.sort((a, b) => a.localeCompare(b));
  unchanged.sort((a, b) => a.localeCompare(b));

  return { added, removed, unchanged, unchangedCount: unchanged.length };
}

export function listDiffFromBridgeDiff(diff: CandidateDiff): ListDiff | null {
  const raw = diff.list_diff;
  if (!raw || typeof raw !== "object") return null;
  const o = raw as Record<string, unknown>;
  return {
    added: Array.isArray(o.added) ? o.added.map(String) : [],
    removed: Array.isArray(o.removed) ? o.removed.map(String) : [],
    unchanged: Array.isArray(o.unchanged) ? o.unchanged.map(String) : [],
    unchangedCount:
      typeof o.unchanged_count === "number"
        ? o.unchanged_count
        : Array.isArray(o.unchanged)
          ? o.unchanged.length
          : 0,
    operation: typeof o.operation === "string" ? o.operation : undefined,
  };
}

export { countImportantTermsInCapture } from "./clipboard-preview.js";

export function listDiffFromProposal(detail: CandidateDetail): ListDiff | null {
  if (!isListDiffSection(detail.proposal.section ?? "")) return null;
  const added = (detail.proposal.items ?? [])
    .map((item) => String(item.name ?? "").trim())
    .filter(Boolean);
  const unchanged = (detail.proposal.already_present ?? [])
    .map((item) => String(item.name ?? "").trim())
    .filter(Boolean);
  return {
    added,
    removed: [],
    unchanged,
    unchangedCount: unchanged.length,
    operation: added.length > 0 ? "list_add" : "list_unchanged",
  };
}

export function importantTermsCardSummary(c: CandidateSummary): string | null {
  if (c.section !== "important_terms") return null;
  const withSummary = c as CandidateSummary & { display_summary?: string | null };
  const summary = withSummary.display_summary?.trim() || c.content_preview?.trim();
  return summary || null;
}

export function proposalItemsForListSection(detail: CandidateDetail): Array<Record<string, unknown>> {
  if (!isListDiffSection(detail.proposal.section ?? "")) {
    return (detail.proposal.items ?? []) as Array<Record<string, unknown>>;
  }
  return (detail.proposal.items ?? []) as Array<Record<string, unknown>>;
}

File: src/sayane/evaluators/list_diff.py
"""List-section diff helpers (important_terms, etc.)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import yaml

from sayane.core.candidate import CandidateProposal

LIST_DIFF_SECTIONS: tuple[str, ...] = (
    "important_terms",
    "tags",
    "keywords",
    "interests",
    "focus_terms",
)


@dataclass(frozen=True)
class ListDiff:
    added: list[str]
    removed: list[str]
    unchanged: list[str]

    @property
    def unchanged_count(self) -> int:
        return len(self.unchanged)

    def to_dict(self) -> dict[str, Any]:
        return {
            "added": self.added,
            "removed": self.removed,
            "unchanged": self.unchanged,
            "unchanged_count": self.unchanged_count,
        }


def diff_string_list(existing: list[str], proposed: list[str]) -> ListDiff:
    """Set-based list diff; order changes are ignored in v1."""
    normalize = lambda s: s.strip()  # noqa: E731

    def norm_key(s: str) -> str:
        return normalize(s).casefold()

    existing_map = {norm_key(v): normalize(v) for v in existing if normalize(v)}
    proposed_map = {norm_key(v): normalize(v) for v in proposed if normalize(v)}

    existing_keys = set(existing_map)
    proposed_keys = set(proposed_map)

    added = [proposed_map[k] for k in sorted(proposed_keys - existing_keys, key=str.lower)]
    removed = [existing_map[k] for k in sorted(existing_keys - proposed_keys, key=str.lower)]
    unchanged = [
        proposed_map[k] for k in sorted(existing_keys & proposed_keys, key=str.lower)
    ]
    return ListDiff(added=added, removed=removed, unchanged=unchanged)


def _normalize_scalar(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, list):
        return ", ".join(_normalize_scalar(v) for v in value if v is not None)
    return str(value).strip()


def parse_yaml_list_section(content: str, section: str) -> list[str]:
    try:
        parsed = yaml.safe_load(content)
    except yaml.YAMLError:
        return []
    if not isinstance(parsed, dict):
        return []
    raw = parsed.get(section)
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for entry in raw:
        text = _normalize_scalar(entry)
        if text:
            out.append(text)
    return out


def important_terms_from_proposal(proposal: CandidateProposal) -> list[str]:
    """All terms referenced in capture classification."""
    names: list[str] = []
    for entry in proposal.items:
        name = (entry.get("name") or "").strip()
        if name:
            names.append(name)
    for entry in proposal.already_present:
        name = (entry.get("name") or "").strip()
        if name:
            names.append(name)
    return names


def list_diff_for_important_terms(proposal: CandidateProposal) -> ListDiff:
    """Diff from proposal classification (added vs already_present)."""
    added = [
        (item.get("name") or "").strip()
        for item in proposal.items
        if (item.get("name") or "").strip()
    ]
    unchanged = [
        (item.get("name") or "").strip()
        for item in proposal.already_present
        if (item.get("name") or "").strip()
    ]
    return ListDiff(added=added, removed=[], unchanged=unchanged)


def important_terms_profile_diff(
    existing: list[str],
    proposed: list[str],
) -> ListDiff:
    """Compare capture list to profile.important_terms only (not other fields)."""
    return diff_string_list(existing, proposed)


def list_diff_operation(diff: ListDiff) -> str:
    if diff.added and diff.removed:
        return "list_update"
    if diff.added:
        return "list_add"
    if diff.removed:
        return "list_remove"
    return "list_unchanged"


def important_terms_display_summary(
    proposal: CandidateProposal,
    *,
    locale: str | None = None,
) -> str:
    diff = list_diff_for_important_terms(proposal)
    if diff.added:
        names = ", ".join(diff.added)
        if locale and str(locale).lower().startswith("ja"):
            return (
                f"important_terms に {len(diff.added)} 件の追加候補があります。"
                f"追加: {names}"
            )
        return f"important_terms: {len(diff.added)} term(s) to add — {names}"
    if diff.unchanged and not diff.added:
        if locale and str(locale).lower().startswith("ja"):
            return f"important_terms: 既存と一致 ({len(diff.unchanged)} 件)"
        return f"important_terms: all {len(diff.unchanged)} term(s) already in profile"
    return proposal.summary or "important_terms"


def important_terms_judge_payload(proposal: CandidateProposal) -> dict[str, Any]:
    diff = list_diff_for_important_terms(proposal)
    return {
        "section": "important_terms",
        "operation": list_diff_operation(diff),
        "added": diff.added,
        "removed": diff.removed,
        "unchanged_count": diff.unchanged_count,
    }
